src/wp-includes/functions.php

```diff
@@ -7984,6 +7984,37 @@ function wp_get_direct_php_update_url() {
 	return $direct_update_url;
 }
 
+/**
+ * Replaces periods with encoded entities in URLs.
+ *
+ * If the URL ends with a period, replace it with %2E to avoid misinterpretation from some email clients.
+ *
+ * @since 5.9.0
+ *
+ * @param string $url The URL to encode.
+ * @return string The new encoded URL.
+ */
+function wp_url_encode_ending_period( $url ) {
+	if ( ! is_string( $url ) ) {
+		return '';
+	}
+
+	$url = trim( $url );
+	if ( '' === $url ) {
+		return '';
+	}
+
+	$url = rawurlencode( $url );
+
+	// If the URL ends in a period, replace it with %2E.
+	if ( str_ends_with( $url, '.' ) ) {
+		$url  = rtrim( $url, '.' );
```

Comment:
```suggestion
		$url  = substr( $url, 0, strlen( $url ) - 1 );
```

Doh, not `rtrim()` (as it removes all of the ending `..`. Nope, use `substr()` instead. Still runs in half the time as `preg_match()`. See the benchmark https://3v4l.org/ZIOQl#v8.0.12.